clang/lib/Sema/SemaStmt.cpp

```diff
@@ -743,6 +743,20 @@ bool Sema::checkMustTailAttr(const Stmt *St, const Attr &MTA) {
     CallerType.Func = CallerDecl->getType()->getAs<FunctionProtoType>();
   }
 
+  if (Context.getTargetInfo().getTriple().isPPC()) {
+    if (Context.getTargetInfo().getTriple().isOSAIX())
+      return Diag(St->getBeginLoc(), diag::err_aix_musttail_unsupported);
+    else if (!Context.getTargetInfo().hasFeature("pcrelative-memops")) {
+      if (Context.getTargetInfo().hasFeature("longcall"))
+        return Diag(St->getBeginLoc(), diag::err_ppc_impossible_musttail) << 0;
+      else if (!CE->getDirectCallee())
+        return Diag(St->getBeginLoc(), diag::err_ppc_impossible_musttail) << 1;
+      else if (isa_and_nonnull<FunctionDecl>(CE->getCalleeDecl()) &&
```

Comment:
Just use the result of getDirectCallee() here.

I'm not sure isDefined() does what you want, for multiple reasons. One, the definition might be later in the file, and two, weak linkage makes "defining" a function a bit strange (e.g. C++ inline functions, or -fsemantic-interposition).

Reply:
Thanks for comment. @efriedma-quic 

Tried the below two cases:
- case 1:
```
int func2();
int external_call2() {
  [[clang::musttail]] return func2();
}

int func2() {
  return 0;
}
```

I think this should be a valid case for tailcall optimization, right? `isDefined()` returns true for the callee, but current `checkMustTailAttr()` treats this as an invalid musttail case. See https://godbolt.org/z/79sYfveds

- case 2:
```
__attribute__((weak)) int func2() {
  return 0;
}
int external_call2() {
  [[clang::musttail]] return func2();
}
```
With this patch, I get:
```
error: 'musttail' attribute for this call is impossible because external calls can not be tail called on PPC
```

I think the diagnose message is right? We have no way to know if there is any strong definition for this weak function, so maybe we should treat it as external calls?

To me, the two cases(if my case 2 can reflect your weak linkage case) are both handled correctly with `isDefined()`, although current `checkMustTailAttr()` does not work well for the functions that have both declaration and definition in the same file.

Reply:
Thanks for comments. @efriedma-quic 

I tried below two cases, 
```
$ cat 1.c
int func2(int i);
int external_call2(int i) {
  [[clang::musttail]] return func2(i);
}

int func2(int i) {
  return 0;
}
$ cat 2.c

__attribute__((weak)) int func2(int i) {
  return 0;
}
int external_call2(int i) {
  [[clang::musttail]] return func2(i);
}
```

- `1.c` is good to compile without this patch, but now it is mistakenly reported as invalid to do tail call because the callee is treated as undefined...
- `2.c` can not be detected by this patch.

Will try to fix this. If you have any idea to fix these two issues in your mind, please let me know!

Reply:
`2.c` can be detected by checking the callee definition is weak.

Reply:
For the weak definitions, you can look at GVALinkage or something like that, yes.

For the ordering issue, you probably need to do the analysis at the end of the translation unit, or something like that.

Reply:
@efriedma-quic sorry for the delay. I've handled the forward declaration case in the new patch. Would you please help to have a look? Thanks very much.